Account for the generic zero-sized payload enum cases.

hjyamauchi opened the pull request:
This fixes an assert failure:

    // At least one payload is non-empty (otherwise this
    // would get laid out as a non-payload enum)
    assert(getNumNonEmptyPayloadCases() > 0);

in TaggedMultiPayloadEnumTypeInfo when you have a generic but zero-sized payload enum cases.

compnerd commented:
@swift-ci please test

tbkka commented:
I'm not sure about this.  My understanding is that the compiler treats all generic payloads as if they were non-zero-sized for the purposes of determining a layout.   Is this right, @slavapestov ? 

If my understanding is correct, then it's probably the assert here that's actually wrong, not the layout determination logic.  I'll take a look.

tbkka commented:
I did a little bit of poking around, and I'm pretty sure that:
* This PR is actually wrong
* (The test case is valuable; we should keep that)
* The assert is wrong and needs to be removed (or even better, commented out with an explanation of why it's wrong)

tbkka reviewed (changes requested):
I believe the changes to TypeLowering.cpp here are all wrong.  

However, the compiler support that this is trying to match is complex; it is quite likely that I don't entirely understand it either.  I'd be happiest if you could come up with a test case that actually shows an incorrect enum reflection for some version of this (not just the failing assertion).  See the tests in validation-tests/Reflection for examples of how to verify enum reflection.

tbkka commented:
With your changes, validation-tests/Reflection/reflect_Enum_MultiPayload_generic7.swift and .../reflect_Enum_MultiPayload_generic8.swift both fail on my local system.  Do you also see this?

tbkka commented:
These cases also failed in CI.  Only for macOS ... I wonder if these tests even run on Linux/Windows?
```
Failed Tests (3):
  Swift-validation(macosx-x86_64) :: Reflection/reflect_Enum_MultiPayload_generic2.swift
  Swift-validation(macosx-x86_64) :: Reflection/reflect_Enum_MultiPayload_generic7.swift
  Swift-validation(macosx-x86_64) :: Reflection/reflect_Enum_MultiPayload_generic8.swift
```

compnerd commented:
https://github.com/apple/swift/blob/main/validation-test/Reflection/reflect_Enum_MultiPayload_generic2.swift#L8
https://github.com/apple/swift/blob/main/validation-test/Reflection/reflect_Enum_MultiPayload_generic7.swift#L8

So, generic2 and generic7 do not run on Windows, generic8 does though.

tbkka commented:
> So, generic2 and generic7 do not run on Windows, generic8 does though.

That looks like a lazy copy/paste.   I don't see any reason those two tests would actually require ObjC interop.  Can someone try removing those lines and see if they work today?

hjyamauchi committed:
Account for the generic zero-sized payload enum cases.

Disable the assert:

    // At least one payload is non-empty (otherwise this                                                                                                                  // would get laid out as a non-payload enum)                                                                                                                          assert(getNumNonEmptyPayloadCases() > 0);

in TaggedMultiPayloadEnumTypeInfo because it fails when you have
generic but zero-sized payload enum cases.

Also remove unnecessary "REQUIRES: objc_interop" lines from the enum
reflection tests.

hjyamauchi commented:
Update this PR to comment out the assert with a comment and to move the test to validation_test. Also removed the objc_interop requirements from several enum reflection tests.

Yes I was able to see the failures with the original PR.
Yes the enum reflection tests didn't need objc_interop.

PTAL.

tbkka reviewed (approved):
Nice!   Thank you so much for sorting through all those tests and getting them to run on Windows and Linux!

tbkka commented:
@swift-ci Please test

hjyamauchi commented:
I was able to repro those failures on Linux. Updated the PR to disable them. PTAL. Can you rerun the CI?

tbkka commented:
@swift-ci Please test

hjyamauchi commented:
The windows test timed out. Can anyone rerun it?

hjyamauchi commented:
@tbkka Would you rerun the windows CI? This is blocking a debugger scenario on windows that we'd like to fix soon.

tbkka commented:
@swift-ci Please test Windows

hjyamauchi commented:
@tbkka Thanks! Do you mind merging this?

tbkka commented:
Done!  Thanks for following through on this!

hjyamauchi commented:
Thank you so much!
